fix(step-functions): assignment during literal can use the wrong values

thantos opened the pull request:
* [x] safe-variable heap assignment approach
* ~~in-order evaluation approach~~

sam-goodwin commented on src/asl.ts:
````diff
@@ -1955,19 +1953,45 @@ export class ASL {
+             * ${x} ${(x = "2")}` // should output `1 2`
+             * ```
+             *
+             * If the heap assignment isn't necessary, we will optimize out the extra assignment later.
````
Does this "later" code exist already? Can you link to where this will be done?

thantos replied:
it does not exist yet, wanted to get everything lined up how I wanted it and then write the optimization. Doing to re-write to execute in order now and will write the optimization after.

thantos replied:
https://github.com/functionless/functionless/issues/415

sam-goodwin replied:
Hmm, can you explain how inefficient the generated ASL will be if we merge this?

thantos replied:
> Hmm, can you explain how inefficient the generated ASL will be if we merge this?

It will work correctly for all cases. 

Each dynamic assignment in an object, array, or template literal will create an extra state and variable.

(This pr is still draft so I can take a crack at the optimization).

sam-goodwin replied:
How may states doe this require?

```ts
{
  constant: "value",
  dynamic: x
}
```

thantos commented:
Attempting another strategy where "dynamic" properties are assigned in order. 

However the optimization of assigning constant values is not as effective as I'd originally hoped. Using this strategy, any json path, conditional, or object literal which uses dynamic values could contain an incorrect value.

ex:
```ts
        let z = "";
        const g = {
          a: z,
          b: (z = "a"),
          c: ((z = "b"), z),
          z,
          t: z === "b",
          o: { z }
        };
```

`a: z` is now correct, being assigned the original value of `z`, but z, t and o are all wrong, being assigned in the initial object.

We could also treat any json path, condition, or literal value with embedded json path as dynamic values, but that would really limit the objects that benifit from this  strategy.

In the name of time time and correctness, will keep the aggressively assigned heap values and look to optimization (#415) to reduce the unnecessary states. 

code: 
```ts
} else if (isObjectLiteralExpr(expr)) {
      return this.evalContext(expr, ({ addState }) => {
        const evaluatedProperties = expr.properties.map((prop) => {
          if (!isPropAssignExpr(prop)) {
            throw new Error(
              `${prop.kindName} is not supported in Amazon States Language`
            );
          }
          if (
            (isComputedPropertyNameExpr(prop.name) &&
              isStringLiteralExpr(prop.name.expr)) ||
            isIdentifier(prop.name) ||
            isStringLiteralExpr(prop.name)
          ) {
            const name = isIdentifier(prop.name)
              ? prop.name.name
              : isStringLiteralExpr(prop.name)
              ? prop.name.value
              : isStringLiteralExpr(prop.name.expr)
              ? prop.name.expr.value
              : undefined;
            if (!name) {
              throw new SynthError(
                ErrorCodes.StepFunctions_property_names_must_be_constant
              );
            }

            return [name, this.eval(prop.expr)] as const;
          } else {
            throw new SynthError(
              ErrorCodes.StepFunctions_property_names_must_be_constant
            );
          }
        });

        const constantProperties = evaluatedProperties.filter(
          (entry): entry is [string, ASLGraph.Output] =>
            ASLGraph.isAslGraphOutput(entry[1])
        );

        const dynamicProperties = evaluatedProperties.filter(
          (
            entry
          ): entry is [
            string,
            ASLGraph.OutputState | ASLGraph.OutputSubState
          ] => !ASLGraph.isAslGraphOutput(entry[1])
        );

        // create an object out of the literal properties
        const constantPayload = constantProperties.reduce(
          (obj, [name, out]) => {
            // if this output is a Condition, need to normalize it.
            const updated = this.normalizeOutputToJsonPathOrLiteralValue(out);
            if (ASLGraph.isStateOrSubState(updated)) {
              addState(updated);
            }
            const updatedOut = ASLGraph.isStateOrSubState(updated)
              ? updated.output
              : updated;
            return {
              value: {
                ...(obj.value as Record<string, any>),
                ...ASLGraph.jsonAssignment(name, updatedOut),
              },
              containsJsonPath:
                obj.containsJsonPath ||
                ASLGraph.isJsonPath(updatedOut) ||
                updatedOut.containsJsonPath,
            };
          },
          { value: {}, containsJsonPath: false }
        );

        // if there are any elements that need to run additional states,
        // first assign the object literal with constant values (if there are any)
        // and then run each of the object assignments in order.
        // Finally, return the heap variable that contains the new object.
        if (dynamicProperties.length > 0) {
          // location where the object will be created
          const objOutput = this.newHeapVariable();

          const dynamicPropertiesStates = ASLGraph.joinSubStates(
            expr,
            constantProperties.length > 0
              ? this.assignValue(expr, constantPayload, objOutput)
              : undefined,
            // run each of the property assignment in order
            ...dynamicProperties.map(([name, states]) => {
              return ASLGraph.joinSubStates(
                states.node,
                states,
                this.assignValue(
                  states.node,
                  states.output,
                  `${objOutput}['${name}']`
                )
              )!;
            })
          )!;

          return {
            ...dynamicPropertiesStates,
            output: { jsonPath: objOutput },
          };
        }

        return constantPayload;
      });
    } 
```

sam-goodwin commented on test/step-function.test.ts:
```diff
@@ -883,7 +883,7 @@ test("put an event bus event", () => {
       Parameters: { Entries: [{ EventBusName: bus.eventBusArn }] },
     },
-    "bus.putEvents"
+    "1__await bus.putEvents"
```
What changed here

thantos replied:
More states, needed to be more specific with the state names.
